### DISCUSSION TITLE
G6 time travel 80/81 series firmware 1.6.5.25

### DISCUSSION THREAD

**Discussion by Foxy7** [Navid200/xDrip#165] [Dexcom] [answered]
Hi @Navid200
I asked on the Gitter chat [link](https://gitter.im/jamorham/xDrip-plus?at=61befaf25a172871a912dcfa), but makes sense to formalise it here.

I'd like to connect xdrip to a sensor that is already embedded and for example 9 days old and time travel back to day 3 manually to avoid the sensor stopping inconveniently on day 10. (or even just time travel 2 hours for a sensor that we put on early to avoid having to wait for warm up, so no down time for looping).

I'm using a rebatteried G6 80 transmitter with firmware 1.6.5.25 which satisfies the requirements for preemptive restart as described [here](https://github.com/Navid200/xDrip/wiki/Preemptive-Restart).

I have two G6 80 TX active all the time running two different systems, one using xdrip+OpenAPS and the other using CamAPS+Bluejay GTS.

**Set Up 1:**
The xdrip+OpenAPS one is set up for engineering mode/extended time travel and works automatically on day 9 to update the TX sensor start to be day 3, this has been running successfully for sometime with no issues. 

xdrip version is latest master 4th Oct 21.

The sensor is started by xdrip in the first instance and connected for the entire duration of the sensor life.

It appears that the time travel process is driven by this part of the code [extended time travel](https://github.com/NightscoutFoundation/xDrip/blob/424f32a37e74c98c4af150eaa8ad774c669b913f/app/src/main/java/com/eveningoutpost/dexdrip/G5Model/Ob1G5StateMachine.java#L1178) triggered via code [day 9](https://github.com/NightscoutFoundation/xDrip/blob/424f32a37e74c98c4af150eaa8ad774c669b913f/app/src/main/java/com/eveningoutpost/dexdrip/G5Model/Ob1G5StateMachine.java#L732). Neither of these sections of code are available via voice commands.

**Set Up 2:**
For the CamAPS+GTS I start the sensor using CamAPS, on sensor day 9 last week i put the CamAPS phone on flight mode and connected to xdrip on another phone.

xdrip version is latest nightly build 8th Dec 21.

It takes at least 15 mins to migrate over to xdrip as the TX has to forget the old bluetooth connection, so agree with these [instructions](https://github.com/Navid200/xDrip/wiki/Session-in-progress).

The TX connected correctly and xdrip read the sensor readings as expected, showing 9.1 days. however xdrip didn't trigger the automatic time travel event despite being 9 days old and having all the correct settings enabled for engineering mode and extended time travel as per set up 1.

![Screenshot_20211219-085423_xDrip+_s1](https://user-images.githubusercontent.com/37312469/147161759-47643c86-ac2d-4a03-93d1-c3c13b644def.jpg)

Consequently I manually ran the voice command ["restart g5 session nearly ended"](https://github.com/NightscoutFoundation/xDrip/blob/3570560268163ee79266945bdd01537572b0e3e6/app/src/main/java/com/eveningoutpost/dexdrip/UtilityModels/VoiceCommands.java#L119), which asked "Needs Initial Calibration", but i don't want to provide calibrations since the sensor was working accurately. Calibrations aren't needed for an automatic extended time travel restart.

![Screenshot_20211219-091732_xDrip+_s2](https://user-images.githubusercontent.com/37312469/147161939-45b73b2a-b45b-4e7a-b7eb-8a008c3dd965.jpg)

Possibly linked to a comment i made previously is that the recent builds including Latest don't give the option to set the time/date of insertion even with correct firmware, which was always asked using last years xdrip builds, but this doesn't explain why the day 9 time travel does work for set up 1. 

So i uninstalled xdrip and loaded up an old version from last year, 25Jul20 and this time starting the sensor and the "Did you insert it today?" message appeared and i set the sensor start date to be three days ago. Alas, this also failed to start correctly and also asked for "Needs Initial Calibration", but the "Classic Status Page" did show the sensor age as 3d 2h old.

![Screenshot_20211219-094214_xDrip+_s3](https://user-images.githubusercontent.com/37312469/147162073-9f31535b-90ae-40a5-9597-77b13ccc1842.jpg)

After this i'd run out of time/patience and just did a normal sensor start and had to wait 2 hours.

I only got the following screenshot of the logs and don't seem to have anything left i can show you now. Next sensor restart in 9 days as just put a new one on.

![Screenshot_20211219-095156_xDrip+_s4](https://user-images.githubusercontent.com/37312469/147162242-2d322fe9-41e3-47a1-90b2-aed2d77537a2.jpg)

I think the issue with the restart might be related to this instance of xdrip not knowing the calibration code to pass to the TX, as the TX needs that to avoid asking for calibration finger tests. Else it's like starting the sensor using code "0000" which is just calibrated to finger tests.

Your suggestion on the chat was to use "set sensor code" (which i've never used before) prior to running the "restart g5 session nearly ended" if i understood you correctly?

Any ideas on areas to investigate?

Thanks for your help, it's very much appreciated.

**Comment by Navid200** [comment]
When xDrip picks up a session in progress, the data on the classic status page with respect to when sensor started is incorrect and disagrees with the information on the G5/G6 status page.
The info on the G5/G6 status page must come from the transmitter, which knows when the sensor started.
The info on the classic status page must just be a timer that xDrip starts when you start the sensor.  In this case, since the sensor was not started by xDrip, I suspect that the timer starts as soon as xDrip identifies there is an active sensor.  So, it assumes that the sensor has just started.

My suspicion is that the preemptive restart code uses the information from the classic status page to decide when day 9 comes.  And as long as you start the sensor with xDrip, there is no inconsistency and everything works fine.
But, if you have started the sensor 9 days ago with another app and just now use xDrip to pick up that session, the classic status page is going to show that the sensor has just started now.  So, it will not trigger a preemptive restart.
But, I cannot emphasize enough that this is just a suspicion.

I don't want to ask too many questions like why do you have two transmitters running at the same time.
But, I have to ask the following question.
Why do you not use xDrip shortly after starting the sensor with another app to recognize the session?
If you do that, you can just close that version of xDrip.  But, in 8.5 days, you can start that xDrip and it will be on day 8.5 and should trigger a preemptive restart on day 9.

**Reply by Foxy7** [reply]
@Navid200  thanks for the reply

Just to clarify, we have two Type 1s in our house hence running two separate loops. describing setup 1 was just to show that time travel does work in the more usual set up using xdrip to start the sensor and automatic restarts on day 9 back to day 3.

I could use xdrip to connect to recognize the session, but would take the best part of an hour to switch from CamAPS to xdrip and back again, and it's not always convenient to do so. 

However i will try this this tonight, as sensor is now 7 days old, so will see if i can link to it to allow xdrip to connect then connect back to CamAPS. I'll report back on success of this and any subsequent auto restart on day 9 using the same approach.


Regarding your comment "..you can just close that version of xDrip." what does closing xdrip mean? 

Note, I've been uninstalling xdrip between uses on my phone since i don't want bluetooth to be on all the time, because somehow just turning off permissions doesn't quite work as i'd like. 

- [ ] possible future feature would be to have an xdrip "flightmode" button within the xdrip app which would stop xdrip calling bluetooth when not desired but maintain all the settings.

**Reply by Navid200** [reply]
I had thought that the two transmitters were being used by one diabetic.  It's clear that I was wrong.  Thanks for clarifying.

Does CamAPS use the mobile slot and not the medical slot of a G6?
If yes, why?  Why have they made it that way and how is everyone using it to control their pump supposed to use their CGM?

If CamAPS uses the same (mobile) slot as xDrip (please answer my question above), and instead of asking them to fix that you are asking us to change xDrip, this is my suggestion:

I take it the only reason you use xDrip is to restart the sensor and nothing else.  Correct me if that's a wrong assumption.
If yes, you can use a different phone for xDrip than for CamAPS.  After starting the sensor with CamAPS, put the CamAPS phone in the microwave or just turn it off.
Turn on the xDrip phone and establish connectivity.  It will take 10-15 minutes.  This will allow the two pages, classic and G5/G6 status, to have very close info with respect to the G6 start time, only 10-20 minutes difference.
After connectivity is established and you have readings in xDrip, shut down the phone, or go to the hardware data source and scroll all the way down and tap on disable collection.
Now, you can turn on the CapAPS phone again.

In 8.5 days, when it's the time for preemptive restart, you will need to again turn off the CapAPS phone and turn on or enable the xDrip phone.
After the preemptive restart is executed, you can disable the xDrip phone again and enable the CamAPS phone.

If this sounds like a cumbersome approach and you are asking yourself "why would I disable the phone that controls the pump for 20 minutes", I feel exactly the same way and that's why I question why they don't do what the t:slim pump dues using the medical slot and leaving the mobile slot alone.

**Reply by Navid200** [reply]
With respect to your suggestion for future, I don't see the reason for it.  You can disable collection, which stops xDrip from operating without changing any settings.

**Reply by Foxy7** [reply]
> disable collection

how do i do this pls?

uncheck the otpions in the "G5/G6 Debug settings" for "Use the OB1 Collector" and "Native Algorithm"?

**Reply by Navid200** [reply]
Please have a look at this:
https://github.com/Navid200/xDrip/wiki/Disable-Collection

Please let me know if it does not answer your question.

**Reply by Foxy7** [reply]
Thanks, i've tried that and bluetooth hasn't turned on again since, seems to have worked. 👍 

Note, i've never seen that before, i have only used the Source Wizard button to change to a non G6 source when i wanted to stop from connecting to the TX.

i searched for the phrase "disable collector" and that doesn't appear:

![image](https://user-images.githubusercontent.com/37312469/147709374-d51bf68b-181a-4fb1-be26-e535e6d590f0.png)


Can i suggest a small change to include "disable collector" in the title and text, because in other places it's called the collector, e.g. "Use the OB1 Collector" and on the Classic Status Page "Restart Collector". I'm happy to do the PR if you'd prefer.

**Reply by Foxy7** [reply]
> Does CamAPS use the mobile slot and not the medical slot of a G6? If yes, why? Why have they made it that way and how is everyone using it to control their pump supposed to use their CGM?

Yes CamAPS uses the mobile slot and you can start/stop sensors directly using CamAPS, no need for any other CMG. But it's not great for monitoring, e.g. a child at school, since the data is only loaded to Diasend. As you know, using Bluejay GTS, you can used that to connect to the TX on the medical slot and disabling the phone collector send the data to nightscout using xdrip, providing using Android 10 at least.


> If CamAPS uses the same (mobile) slot as xDrip (please answer my question above), and instead of asking them to fix that you are asking us to change xDrip, this is my suggestion:

Not quite, my preference would be to be able to select the slot within xdrip, as we can with Bluejay or OpenAPS/Logger.


> I take it the only reason you use xDrip is to restart the sensor and nothing else. Correct me if that's a wrong assumption. If yes, you can use a different phone for xDrip than for CamAPS. After starting the sensor with CamAPS, put the CamAPS phone in the microwave or just turn it off. Turn on the xDrip phone and establish connectivity. It will take 10-15 minutes. This will allow the two pages, classic and G5/G6 status, to have very close info with respect to the G6 start time, only 10-20 minutes difference. After connectivity is established and you have readings in xDrip, shut down the phone, or go to the hardware data source and scroll all the way down and tap on disable collection. Now, you can turn on the CapAPS phone again.
> 
> In 8.5 days, when it's the time for preemptive restart, you will need to again turn off the CapAPS phone and turn on or enable the xDrip phone. After the preemptive restart is executed, you can disable the xDrip phone again and enable the CamAPS phone.

Thanks, i now understand what you meant in your first message, we need the two systems to have the same time stamps so need to connect within a few minutes of starting. i read this as, really i should always start the new sensor with xdrip, then switch to CamAPS once the 2 hours are up and set "disable collection" as hardware.

> If this sounds like a cumbersome approach and you are asking yourself "why would I disable the phone that controls the pump for 20 minutes", I feel exactly the same way and that's why I question why they don't do what the t:slim pump dues using the medical slot and leaving the mobile slot alone.

Agree, this would make more sense.

**Reply by Navid200** [reply]
In the title and text of what?

**Reply by Navid200** [reply]
With respect to the slot, I have already brought this up.
The concern is that even if we give the capability to xDrip to use the medical device, as long as CamAPS is on the same phone, it will be using the same pairing, which could still not work.

**Reply by Foxy7** [reply]
> In the title and text of what?

In the link you sent to your wiki, for some reason a search for Disable Collector doesn't direct me to this page: https://github.com/Navid200/xDrip/wiki/Disable-Collection

**Reply by Navid200** [reply]
Now I understand.
The search function is lacking.
Thanks so much for bringing this up.  
I have to come up with a search engine.

Thanks so much.  It will be done.

**Comment by Foxy7** [comment]
I've tested this out this evening, i put the CamAPS phone on flightmode was able to link xdrip to the sensor started via CamAPS and then switched it back to CamAPS, took about an 50 mins in total.

xdrip gave the following results:

```
Classic Status Page 
Sensor Start: 29/12/2021 18:26 (0d 3h)

G5/G6 Status
Sensor Status: OK 7.1 days
```

As you expected, the Classic Status is out of sync, which you think is driving the preemptive restart task. I'll have a look at the code tomorrow, but i'm new to java so will take me some time to get through it.

**Reply by Navid200** [reply]
It should make no difference if you have xDrip connect shortly after you start the sensor with CamAPS.  If you do that, the difference between the two will be at the most one hour, which should not cause a problem with preemptive restart in 9 days.

**Reply by Foxy7** [reply]
Agree, providing i link up at the start of the new sensor. 

But having linked to xdrip last night on day 7.1, Classic Status Page shows only 3h old and you think this age drives the preemptive restart task. I would think the TX sensor age is the correct one to use to do the preemptive restart from, since that is the one that stops when it gets to day 10.

**Reply by Navid200** [reply]
I have never looked at the xDrip code to find out why this happens.  But, as I have noticed the discrepancy between the two pages, I suspect that's the reason.
And if that is, a change to xDrip can fix it.
However, making a change is not enough to get it approved.

If I submit a request for change, the first question I will be asked is "why do we need this".
After I explain the whole thing, I will be asked "there must be a workaround".  I will then tell them what the workaround is.  And that will be the end of it.  You can see how many open PRs there are that are not merged.

There is a price to pay for every change we make to xDrip.
So, we always question every PR intentionally to make sure it is absolutely necessary.  I om one of the ones who believes we should only make changes that re absolutely necessary.
Changes that re only needed for a small percentage of xDrip users take a backseat unfortunately.

I know this sound unfair.
But, even testing this is going to be a challenge.  None of the developers has an old G6.  So, how can we even test it?

Please don't forget that all developers are volunteers.

If I understand why the workaround I have explained does not work, I will be in a better position to get this done.
But, your descriptions of what you have tried often leaves the most crucial point out.
I still don't understand if you have tried what I proposed or not.
I understand you have made some experiments.  But, I wish you would tell me "I did exactly what you suggested;  but, it did not do what you said it would".

My expectation is that if xDrip connects to a sensor shortly after it started, xDrip will remember it.  If now xDrip stops communication and reestablished in about 9 days, it should still show the correct start time coming from its memory.  If I am wrong about this, my suggested workaround will not work.  But, have you done this?

**Reply by Foxy7** [reply]
Hi @Navid200

Thanks for you comments, please know that i can see how many demands are on you and the other volunteers when i read all the chats and try to follow these discussions, so all of your efforts are very very much appreciated by me and everyone of us who use xdrip.

I'll try the workaround on the next sensor change, which is in the next few days, so starting using my phone's xdrip and then Disable-Collection and transferring across to CamAPS, so hopefully we can make use of the preemptive restart on day 9.

I'll let you know how it goes.

Thanks again and happy new year!

**Reply by Foxy7** [reply]
Hi @Navid200 good news, success!

Turns out you were right, the key trick was to set the calibration code using `set sensor code` **before** doing the restart.

Sensor was due to run out tonight and I connected back to xDrip just before lunch to see if i could get it to restart, first attempt failed as no calibration code set prior to preemptive restart, but after running the voice commands `set sensor code` then `restart G5 session` it worked correctly and set back to `3d 1h`.

For today, i used CamAPS and xDrip on two separate phones:

- After reconnection xDrip automatically did a preemptive restart since the TX days = 9, despite the "Classic Status Page" showing only 2d 17h since i'd only connected to the TX for the first time on Weds evening (note i'd started the sensor with CamAPS, not xDrip)
- xDrip reported `Needs Initial Calibration`
- Error log has the line `Cannot reprocess start G6 sensor as calibration code not set!` This is a slightly different error from the one in my original post
- i set the sensor code and then `restart G5 session`, on the next reading it G5/G6 Status had `OK 3.1 days` and readings were showing again on the graph
- waited for the next reading to double check it was working properly
- Disabled Collection
- Linked back to CamAPS which picked up the TX about 15 mins later showing `3d 2h 33min`

The whole process took about 50 mins, including me getting it wrong the first time, so would hope I can get down to <30 min with the instructions below.

I have screen shots of all of this if of interest, but don't think it adds much given our discussion so far.

Cheers

**Comment by Navid200** [comment]
I was thinking about this whole concept again.
If you cannot start the sensor with xDrip, how can you restart it with xDrip?

I know that t:slim needs to start the sensor.  The reason is that if you start with xDrip, t:slim cannot pick up the session.
So, if the issue is the same with CamAPS, then if you restart with xDrip, doesn't CamAPS still expect it to end when it was going to?

**Reply by jonprogers** [reply]
Morning,
I'm a t:slim user and always start and stop my sensors with xDrip, t:slim picks the session up no problem at all as long as it already has the correct tx id.
This is for UK supplied t:slim running software version:- Moonlight v7.4
Only slight wrinkle is that it takes an extra 5 minute cycle for the pump to pick up the session as with firefly transmitters the comms with the tx is completed first for the medical slot then it does the mobile slot but for earlier devices I believe they were both instigated together (long time since I used a pre firefly tx!!).

**Reply by Foxy7** [reply]
Thanks @jonprogers  that fits with what i've seen using CamAPS.

we can connect to a sensor started via xdrip simply enough, but agree that on a new pairing request it drops the other slot for 5 mins. e.g. last night switching from CamAPS to xdrip on mobile slot, caused the Bluejay GTS (using medical device slot) to miss a reading when xdrip paired. The bluejay missed the next reading too, so restarted it and it connected correctly on the next reading.

**Reply by Navid200** [reply]
If CamAPS can pick up a session started by xDrip, why don't you always start with xDrip then?
If you do, the two system status pages of xDrip will always be in agreements.  When the time comes for the preemptive restart, you will need to disable CampAPS and enable xDrip until the restart is completed and switch back to CampAPS again.

**Reply by Navid200** [reply]
@jonprogers I had been told that sensor had to be started on t:slim.  Has there been a recent change?  Or, was I misinformed and this could always be done?

**Reply by jonprogers** [reply]
I've been on the t:slim since August 2019 and have always started and stopped sensors successfully using xDrip without issue. This worked in this way for both BasalIQ and the newer ControlIQ versions of the t:slim software so my initial reaction is that you were misinformed. The only difference MAY be that I am a UK user and there is a slight difference in that I share pump data with my endo through Diasend rather than the t:connect software - not sure if that means there is a slightly different software variant in the UK that may have differing quirks...

**Reply by Foxy7** [reply]
> If CamAPS can pick up a session started by xDrip, why don't you always start with xDrip then? If you do, the two system status pages of xDrip will always be in agreements. When the time comes for the preemptive restart, you will need to disable CampAPS and enable xDrip until the restart is completed and switch back to CampAPS again.

Agree, I think this is the only work around available at present, but i didn't know this was necessary until today, hence why i've not tried it before.

In the immediate time frame, it does mean i won't be able to restart the the current sensor which hits Day 9 tomorrow night via xdrip, unless i can restart only a couple of days back? but i still have the issue that xdrip doesn't give me the option to back date the sensor insertion, perhaps these two things are linked (seems like they are).

**Comment by Foxy7** [comment, marked as answer]
So going forward the answer for me is:

**Just as a reminder, this only works for G6 80/81 series firmware 1.6.5.25**

Health warning, this restart is to give me the convenience of when we choose to put on a new sensor, which is usually around day 10, but avoids the sensor from stopping when inconvenient e.g. overnight or whilst away from home.

Assumes xDrip is set up on the same phone as CamAPS and correctly enabled for preemptive restarts.

Also assumes that the G6 sensor is on day 9.

**CamAPS**

1. Check when the last connection to the sensor was, if within the next 2 mins best to wait for that reading to give enough time to switch (although shouldn't be an issue if you don't, just saves missing one)
2. Disable CamAPS for a short time by setting the transmitter number to be "810000" or something that isn't the actual transmitter id

**xDrip**

3. Set Hardware Data Source = G5/G6 Transmitter
4. Check TX id is correct on G5/G6 Status page
5. Use voice command `set sensor code` and type in the initial four digit code, e.g. `5937` best to take a photo of current sensor code every time you start a new one. Check the correct Calibration Code is visible on the G5/G6 Status page
6. If the sensor age is >9 days then xDrip will preemptively restart the sensor, however if it connects and tries to restart before you managed to set the senor code (the Event Log will have the following highlighted in red `Cannot reprocess start G6 sensor as calibration code not set!`) then you need to do the following:
  a. Use voice command `set sensor code` and type in the initial four digit code
  b. Check the correct Calibration Code is visible on the G5/G6 Status page
  c. Use voice command `restart G5 session` to time travel back to day 3
7. If successful the next reading will show a glucose value and on the G5/G6 Status will show `Sensor Status : OK 3.1 days`
8. Set Hardware Data Source = [Disable Collection](https://github.com/Navid200/xDrip/wiki/Disable-Collection)


**CamAPS**

9. Set transmitter back to actual serial number
10. The glucose will be visible on the next connection, it will missed the ones in between, check "Since insertion" time should show something like `3d 2h 30min`

**Reply by Foxy7** [reply]
@Navid200

I the process above last night as the sensor got to day 9 and it was successful, took 15 minutes to complete the cycle from CamAPS to xDrip restart and back to CamAPS, with CamAPS only missed two readings, i think the absolute quickest this can be done, so gvery happy. Note xDrip captured both those readings, so no actual lost data.

I installed xDrip on the same phone as CamAPS earlier in the afternoon (Android 8), took <5 mins to set up and added the sensor code. I put the phone on flight mode as both CamAPS and xDrip would be searching for the TX at the same time, before setting xDrip `Hardware Data Source = Disable Collection`.

Then after the sensor had passed 9 days, i followed the instructions above. 

It's better having them on the same phone as no need to have to wait for a pairing request, it's immediate on the next connection.

You can close this discussion now, answer marked above.

Thanks  for your help.

**Reply by Foxy7** [reply]
i've also updated the steps in the Answer above to simplify.
